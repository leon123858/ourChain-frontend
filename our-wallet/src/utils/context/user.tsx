import {createContext, ReactNode, useEffect, useState} from 'react';
import {onAuthStateChanged, signOut} from "../userApiWrapper/firebaseAuth.ts";
import app from '../firebase'
import {Auth, getAuth, NextOrObserver, User} from "firebase/auth";

const UserContext = createContext({
    isLogin: false,
    email: '',
    handleLogout: () => Promise.resolve(),
});

function UserContextProvider({children, authStateSubscribe = onAuthStateChanged, signOutBeforeHook = signOut}: {
    children: ReactNode,
    authStateSubscribe?: (auth: Auth | undefined, callback: NextOrObserver<User | null>) => void,
    signOutBeforeHook?: (auth?: Auth) => Promise<void>,
}) {
    const [isLogin, setLogin] = useState(false);
    const [email, setEmail] = useState('');

    useEffect(() => {
        authStateSubscribe(getAuth(app), (user) => {
            if (user) {
                setLogin(true);
                setEmail(user.email || "");
            } else {
                setLogin(false);
            }
        })
    }, []);

    const handleLogout = async () => {
        // firebase logout
        await signOutBeforeHook(getAuth(app))
        // state logout
        setLogin(false);
    }

    const defaultValue = {
        isLogin,
        handleLogout,
        email,
    };

    return (
        <UserContext.Provider value={defaultValue}>
            {children}
        </UserContext.Provider>
    );
}

export {UserContextProvider, UserContext};