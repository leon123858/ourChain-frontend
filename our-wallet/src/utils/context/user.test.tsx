import renderer from 'react-test-renderer'
import {describe, expect, test} from "vitest";
import {UserContext, UserContextProvider} from "./user.tsx";
import {useContext} from "react";


describe('context user', () => {
    const MockApp = () => {
        const {isLogin, handleLogout, email} = useContext(UserContext);
        return (<>
                {isLogin || <span>notLogin</span>}
                <h1>email: {email}</h1>
                <button onClick={async () => {
                    await handleLogout();
                }}>logout
                </button>
            </>
        )
    }
    test('try context', async () => {
        let spyTriggerTime = 0;
        const component = renderer.create(
            <UserContextProvider authStateSubscribe={() => {
                // base on use effect, is async call
                spyTriggerTime += 2;
            }} signOutBeforeHook={async () => {
                spyTriggerTime += 1;
            }}>
                <MockApp></MockApp>
            </UserContextProvider>,
        )

        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
        expect(tree).toBeDefined()

        const instance = component.root
        expect(instance).toBeDefined()

        const btn = instance.findByType('button')
        expect(btn).toBeDefined()
        btn.props.onClick()

        expect(spyTriggerTime).toBe(1)

        const h1 = instance.findByType('h1')
        expect(h1).toBeDefined()
        expect(h1.children).toEqual(['email: '])

        const span = instance.findByType('span')
        expect(span).toBeDefined()
        expect(span.children).toEqual(['notLogin'])
    })
})
