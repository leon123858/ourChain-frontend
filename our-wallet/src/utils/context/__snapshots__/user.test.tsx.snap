// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`context user > try context 1`] = `
[
  <span>
    notLogin
  </span>,
  <h1>
    email: 
  </h1>,
  <button
    onClick={[Function]}
  >
    logout
  </button>,
]
`;
